src/bin/winkernel/nmrprovider.c

```diff
@@ -0,0 +1,133 @@
+    _Out_ const void **ProviderDispatch
+)
+{
+    UNREFERENCED_PARAMETER(ClientBindingContext);
+    UNREFERENCED_PARAMETER(ProviderContext);
+    UNREFERENCED_PARAMETER(ClientDispatch);
+
+    *ProviderBindingContext = NmrBindingHandle;
+    *ProviderDispatch = &MsQuicNmrDispatch;
+
+    QuicTraceLogInfo(
+        ProviderAttachClient,
+        "[ nmr][%p] Client attached Ver %hu Size %hu Number %u ModuleID { %x-%x-%x-%x }",
+        NmrBindingHandle,
+        ClientRegistrationInstance->Version,
+        ClientRegistrationInstance->Size,
+        ClientRegistrationInstance->Number,
+        ClientRegistrationInstance->ModuleId->Guid.Data1,
+        ClientRegistrationInstance->ModuleId->Guid.Data2,
+        ClientRegistrationInstance->ModuleId->Guid.Data3,
+        *((uint64_t*)ClientRegistrationInstance->ModuleId->Guid.Data));
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+MsQuicNmrProviderDetachClient(
+    _In_ void *ProviderBindingContext
+    )
+{
+    QuicTraceLogInfo(
+        ProviderDetachClient,
+        "[ nmr][%p] Client detached",
+        ProviderBindingContext);
+    return STATUS_SUCCESS;
+}
+
+_No_competing_thread_
+_IRQL_requires_max_(PASSIVE_LEVEL)
+NTSTATUS
+MsQuicRegisterNmrProvider(
+    void
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+    NPI_REGISTRATION_INSTANCE *ProviderRegistrationInstance;
+
+    NmrProvider.ModuleId.Length = sizeof(NmrProvider.ModuleId);
+    NmrProvider.ModuleId.Type = MIT_GUID;
+    NmrProvider.ModuleId.Guid = MSQUIC_MODULE_ID;
+
+    NmrProvider.NpiProviderCharacteristics.Version = 0;
+    NmrProvider.NpiProviderCharacteristics.Length = sizeof(NmrProvider.NpiProviderCharacteristics);
+    NmrProvider.NpiProviderCharacteristics.ProviderAttachClient = MsQuicNmrProviderAttachClient;
+    NmrProvider.NpiProviderCharacteristics.ProviderDetachClient = MsQuicNmrProviderDetachClient;
+
+    ProviderRegistrationInstance = &NmrProvider.NpiProviderCharacteristics.ProviderRegistrationInstance;
+    ProviderRegistrationInstance->Version = 0;
+    ProviderRegistrationInstance->Size = sizeof(*ProviderRegistrationInstance);
+    ProviderRegistrationInstance->NpiId = &MSQUIC_NPI_ID;
+    ProviderRegistrationInstance->ModuleId = &NmrProvider.ModuleId;
+
+    Status =
+        NmrRegisterProvider(
+            &NmrProvider.NpiProviderCharacteristics,
+            &NmrProvider,
+            &NmrProvider.NmrProviderHandle);
+    if (!NT_SUCCESS(Status)) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "NmrRegisterProvider");
+        goto Exit;
+    }
+
+Exit:
+    return Status;
+}
+
+void
+MsQuicDeregisterNmrProvider(
+    void
+    )
+{
+    if (NmrProvider.NmrProviderHandle != NULL) {
+        NTSTATUS Status = NmrDeregisterProvider(NmrProvider.NmrProviderHandle);
+        CXPLAT_FRE_ASSERTMSG(Status == STATUS_PENDING, "deregistration failed");
+        NmrWaitForProviderDeregisterComplete(NmrProvider.NmrProviderHandle);
```

Comment:
I assume one of these functions is a blocking call, waiting on clients to detach completely too? Should we explicitly try to kill any active connections if we want to detach (i.e. service shutdown) in the middle of things?

Reply:
`NmrWaitForProviderDeregisterComplete` is blocking and returns when the client's detach callback completes inline or asynchronously.

When the provider initiates the deregistration, the client gets a detach callback. The client should kill active connections and any other clean-up work if needed and wait for in-progress operations to complete. The provider should not need to do anything except signaling the client, which is done by NMR firing the client detach callback.

Reply:
Ok. Sounds reasonable.